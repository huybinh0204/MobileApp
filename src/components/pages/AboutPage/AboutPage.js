import Clipboard from '@react-native-community/clipboard';
import React, { useState } from 'react';
import { TouchableOpacity } from 'react-native';
import ENV from 'react-native-config';
import { getBuildNumber, getUniqueId, getVersion } from 'react-native-device-info';
import { IconSvg, Toast } from '_components/atoms';
import { SecondaryScreenLayout } from '_components/organisms';
import { ICONS, TOAST_TYPES } from '_constants';
import strings from '_localization';
import { normalize } from '_utilities/screen';
import { Content, logoStyles, Text } from './AboutPage.styles';

const AboutPage = () => {
  const appVersion = getVersion();
  const buildNumber = getBuildNumber();
  const deviceId = getUniqueId();
  const environment = ENV.BUILD_VARIANT;

  const [toastContent, setToastContent] = useState('');
  const [isToastVisible, setIsToastVisible] = useState(false);

  const copyData = (data) => () => {
    Clipboard.setString(data);
    setIsToastVisible(true);
    setToastContent(`${data} ${strings.about.savedToClipboard}`);
  };

  const closeToast = () => {
    setIsToastVisible(false);
  };

  return (
    <SecondaryScreenLayout testID="AboutPage" title={strings.settings.about}>
      <Content>
        <IconSvg
          icon={ICONS.logoVertical}
          style={logoStyles}
          width={normalize(150)}
          height={normalize(150)}
        />
        <TouchableOpacity accessibilityLabel="Copy app version" onLongPress={copyData(appVersion)}>
          <Text>
            {strings.about.version}: {appVersion}
          </Text>
        </TouchableOpacity>
        <TouchableOpacity
          accessibilityLabel="Copy build number"
          onLongPress={copyData(buildNumber)}
        >
          <Text>
            {strings.about.build}: {buildNumber}
          </Text>
        </TouchableOpacity>
        {environment !== 'PROD' && (
          <TouchableOpacity accessibilityLabel="Copy device ID" onLongPress={copyData(deviceId)}>
            <Text numberOfLines={1}>
              {strings.about.deviceId}: {deviceId}
            </Text>
          </TouchableOpacity>
        )}
      </Content>
      <Toast
        type={TOAST_TYPES.INFO}
        content={toastContent}
        header={strings.about.copied}
        onClose={closeToast}
        show={isToastVisible}
        testID="aboutToast"
      />
    </SecondaryScreenLayout>
  );
};

export default AboutPage;
