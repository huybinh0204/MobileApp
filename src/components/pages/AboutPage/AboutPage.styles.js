import styled, { css } from '@emotion/native';
import { BaseText } from '_components/atoms';

export const Content = styled.View(({ theme }) => ({
  flex: 1,
  alignItems: 'center',
  justifyContent: 'center',
  paddingHorizontal: theme.spacing.m,
}));

export const logoStyles = css({
  marginBottom: 30,
});

export const Text = styled(BaseText)(({ theme }) => ({
  ...theme.typography.mainContentRegular,
  color: theme.colors.beta400,
  textAlign: 'center',
  marginBottom: theme.spacing.xs,
}));
