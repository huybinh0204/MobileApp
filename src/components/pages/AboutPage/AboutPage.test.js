import Clipboard from '@react-native-community/clipboard';
import { fireEvent, render, waitFor } from '@testing-library/react-native';
import React from 'react';
import ComponentWithProviders from '_test/mocks/ComponentWithProviders';
import AboutPage from './AboutPage';

describe('AboutPage', () => {
  it('should match the snapshot', async () => {
    const { toJSON } = render(<ComponentWithProviders component={AboutPage} />);

    await waitFor(() => {
      expect(toJSON()).toMatchSnapshot();
    });
  });

  it('should copy the account number', async () => {
    jest.spyOn(Clipboard, 'setString');

    const { getByAccessibilityLabel } = render(<ComponentWithProviders component={AboutPage} />);
    const appVersion = getByAccessibilityLabel('Copy app version');
    const buildNumber = getByAccessibilityLabel('Copy build number');
    const deviceID = getByAccessibilityLabel('Copy device ID');

    await waitFor(() => {
      fireEvent(appVersion, 'onLongPress');
      expect(Clipboard.setString).toBeCalledWith('appVersion');

      fireEvent(buildNumber, 'onLongPress');
      expect(Clipboard.setString).toBeCalledWith('buildNumber');

      fireEvent(deviceID, 'onLongPress');
      expect(Clipboard.setString).toBeCalledWith('deviceId');
    });
  });
});
