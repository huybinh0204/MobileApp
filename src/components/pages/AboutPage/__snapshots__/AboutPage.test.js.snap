// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AboutPage should match the snapshot 1`] = `
<View
  style={
    Object {
      "flex": 1,
    }
  }
>
  <RNCSafeAreaProvider
    onInsetsChange={[Function]}
    style={
      Array [
        Object {
          "flex": 1,
        },
        undefined,
      ]
    }
  >
    <View
      style={
        Array [
          Object {
            "backgroundColor": "rgb(242, 242, 242)",
            "flex": 1,
          },
          undefined,
        ]
      }
    >
      <RNSScreenContainer
        onLayout={[Function]}
        style={
          Object {
            "flex": 1,
          }
        }
      >
        <RNSScreen
          activityState={2}
          collapsable={false}
          forwardedRef={[Function]}
          pointerEvents="box-none"
          style={
            Object {
              "bottom": 0,
              "left": 0,
              "position": "absolute",
              "right": 0,
              "top": 0,
            }
          }
        >
          <View
            collapsable={false}
            style={
              Object {
                "opacity": 1,
              }
            }
          />
          <View
            accessibilityElementsHidden={false}
            closing={false}
            gestureVelocityImpact={0.3}
            importantForAccessibility="auto"
            onClose={[Function]}
            onGestureBegin={[Function]}
            onGestureCanceled={[Function]}
            onGestureEnd={[Function]}
            onOpen={[Function]}
            onTransition={[Function]}
            pointerEvents="box-none"
            style={
              Array [
                Object {
                  "display": "flex",
                  "overflow": undefined,
                },
                Object {
                  "bottom": 0,
                  "left": 0,
                  "position": "absolute",
                  "right": 0,
                  "top": 0,
                },
              ]
            }
            transitionSpec={
              Object {
                "close": Object {
                  "animation": "spring",
                  "config": Object {
                    "damping": 500,
                    "mass": 3,
                    "overshootClamping": true,
                    "restDisplacementThreshold": 10,
                    "restSpeedThreshold": 10,
                    "stiffness": 1000,
                  },
                },
                "open": Object {
                  "animation": "spring",
                  "config": Object {
                    "damping": 500,
                    "mass": 3,
                    "overshootClamping": true,
                    "restDisplacementThreshold": 10,
                    "restSpeedThreshold": 10,
                    "stiffness": 1000,
                  },
                },
              }
            }
          >
            <View
              collapsable={false}
              pointerEvents="box-none"
              style={
                Object {
                  "flex": 1,
                  "marginTop": 0,
                }
              }
            >
              <View
                collapsable={false}
                forwardedRef={[Function]}
                needsOffscreenAlphaCompositing={false}
                onGestureHandlerEvent={[Function]}
                onGestureHandlerStateChange={[Function]}
                style={
                  Object {
                    "flex": 1,
                    "transform": Array [
                      Object {
                        "translateX": 0,
                      },
                      Object {
                        "translateX": 0,
                      },
                    ],
                  }
                }
              >
                <View
                  style={
                    Array [
                      Object {
                        "flex": 1,
                        "overflow": "hidden",
                      },
                      Array [
                        Object {
                          "backgroundColor": "rgb(242, 242, 242)",
                        },
                        undefined,
                      ],
                    ]
                  }
                >
                  <View
                    style={
                      Object {
                        "flex": 1,
                        "flexDirection": "column-reverse",
                      }
                    }
                  >
                    <View
                      style={
                        Object {
                          "flex": 1,
                        }
                      }
                    >
                      <RCTSafeAreaView
                        backgroundColor={null}
                        emulateUnlessSupported={true}
                        style={
                          Array [
                            Object {
                              "backgroundColor": "#FAFAFA",
                              "flex": 1,
                            },
                            undefined,
                          ]
                        }
                        testID="AboutPage"
                      >
                        <View
                          style={
                            Array [
                              Object {
                                "alignItems": "center",
                                "height": 108.5,
                                "justifyContent": "center",
                              },
                              undefined,
                            ]
                          }
                        >
                          <View
                            accessible={true}
                            collapsable={false}
                            focusable={true}
                            onClick={[Function]}
                            onResponderGrant={[Function]}
                            onResponderMove={[Function]}
                            onResponderRelease={[Function]}
                            onResponderTerminate={[Function]}
                            onResponderTerminationRequest={[Function]}
                            onStartShouldSetResponder={[Function]}
                            style={
                              Object {
                                "left": 36,
                                "opacity": 1,
                                "position": "absolute",
                                "zIndex": 10,
                              }
                            }
                            testID="BackButton"
                          >
                            <SvgMock
                              height={45.5}
                              width={45.5}
                            />
                          </View>
                          <Text
                            allowFontScaling={false}
                            style={
                              Array [
                                Object {
                                  "color": "#1F1D25",
                                  "fontFamily": "Poppins-SemiBold",
                                  "fontSize": 29,
                                  "lineHeight": 40,
                                },
                                undefined,
                              ]
                            }
                          >
                            About
                          </Text>
                        </View>
                        <View
                          onLayout={[Function]}
                          style={
                            Array [
                              Array [
                                Object {
                                  "flex": 1,
                                },
                                undefined,
                              ],
                              Object {
                                "paddingBottom": 0,
                              },
                            ]
                          }
                        >
                          <View
                            style={
                              Array [
                                Object {
                                  "alignItems": "center",
                                  "flex": 1,
                                  "justifyContent": "center",
                                  "paddingHorizontal": 58,
                                },
                                undefined,
                              ]
                            }
                          >
                            <SvgMock
                              height={271.5}
                              style={
                                Object {
                                  "marginBottom": 30,
                                }
                              }
                              width={271.5}
                            />
                            <View
                              accessibilityLabel="Copy app version"
                              accessible={true}
                              collapsable={false}
                              focusable={false}
                              onClick={[Function]}
                              onResponderGrant={[Function]}
                              onResponderMove={[Function]}
                              onResponderRelease={[Function]}
                              onResponderTerminate={[Function]}
                              onResponderTerminationRequest={[Function]}
                              onStartShouldSetResponder={[Function]}
                              style={
                                Object {
                                  "opacity": 1,
                                }
                              }
                            >
                              <Text
                                allowFontScaling={false}
                                style={
                                  Array [
                                    Object {
                                      "color": "#7F8892",
                                      "fontFamily": "Poppins",
                                      "fontSize": 29,
                                      "lineHeight": 40,
                                      "marginBottom": 14.5,
                                      "textAlign": "center",
                                    },
                                    undefined,
                                  ]
                                }
                              >
                                Version
                                : 
                                appVersion
                              </Text>
                            </View>
                            <View
                              accessibilityLabel="Copy build number"
                              accessible={true}
                              collapsable={false}
                              focusable={false}
                              onClick={[Function]}
                              onResponderGrant={[Function]}
                              onResponderMove={[Function]}
                              onResponderRelease={[Function]}
                              onResponderTerminate={[Function]}
                              onResponderTerminationRequest={[Function]}
                              onStartShouldSetResponder={[Function]}
                              style={
                                Object {
                                  "opacity": 1,
                                }
                              }
                            >
                              <Text
                                allowFontScaling={false}
                                style={
                                  Array [
                                    Object {
                                      "color": "#7F8892",
                                      "fontFamily": "Poppins",
                                      "fontSize": 29,
                                      "lineHeight": 40,
                                      "marginBottom": 14.5,
                                      "textAlign": "center",
                                    },
                                    undefined,
                                  ]
                                }
                              >
                                Build
                                : 
                                buildNumber
                              </Text>
                            </View>
                            <View
                              accessibilityLabel="Copy device ID"
                              accessible={true}
                              collapsable={false}
                              focusable={false}
                              onClick={[Function]}
                              onResponderGrant={[Function]}
                              onResponderMove={[Function]}
                              onResponderRelease={[Function]}
                              onResponderTerminate={[Function]}
                              onResponderTerminationRequest={[Function]}
                              onStartShouldSetResponder={[Function]}
                              style={
                                Object {
                                  "opacity": 1,
                                }
                              }
                            >
                              <Text
                                allowFontScaling={false}
                                numberOfLines={1}
                                style={
                                  Array [
                                    Object {
                                      "color": "#7F8892",
                                      "fontFamily": "Poppins",
                                      "fontSize": 29,
                                      "lineHeight": 40,
                                      "marginBottom": 14.5,
                                      "textAlign": "center",
                                    },
                                    undefined,
                                  ]
                                }
                              >
                                Device Id
                                : 
                                deviceId
                              </Text>
                            </View>
                          </View>
                          <View
                            onLayout={[Function]}
                            paddingHorizontal={null}
                            style={
                              Array [
                                Object {
                                  "alignItems": "center",
                                  "bottom": 0,
                                  "paddingHorizontal": 58,
                                  "position": "absolute",
                                  "width": "100%",
                                },
                                Object {
                                  "opacity": 0,
                                  "transform": Array [
                                    Object {
                                      "translateY": AnimatedValue {
                                        " __value": 0,
                                      },
                                    },
                                  ],
                                },
                              ]
                            }
                            testID="aboutToast"
                          >
                            <View
                              style={
                                Array [
                                  Object {
                                    "alignSelf": "stretch",
                                    "backgroundColor": "#FFFFFF",
                                    "borderColor": "#F2F3F4",
                                    "borderRadius": 10,
                                    "borderWidth": 0,
                                    "padding": 29,
                                    "shadowColor": "#362E60",
                                    "shadowOffset": Object {
                                      "height": 7,
                                      "width": 0,
                                    },
                                    "shadowOpacity": 0.1,
                                    "shadowRadius": 10,
                                  },
                                  Array [
                                    Object {
                                      "backgroundColor": "#3D3397",
                                      "borderColor": "#3D3397",
                                      "borderWidth": 1,
                                      "width": "100%",
                                    },
                                    undefined,
                                  ],
                                ]
                              }
                              type="info"
                            >
                              <View
                                style={
                                  Array [
                                    Object {
                                      "alignItems": "flex-start",
                                      "flexDirection": "row",
                                      "justifyContent": "space-between",
                                    },
                                    undefined,
                                  ]
                                }
                                testID="toastHeader"
                              >
                                <SvgMock
                                  height={36}
                                  width={36}
                                />
                                <Text
                                  allowFontScaling={false}
                                  numberOfLines={2}
                                  style={
                                    Array [
                                      Object {
                                        "color": "#FFFFFF",
                                        "flex": 1,
                                        "fontFamily": "Poppins-SemiBold",
                                        "fontSize": 21.5,
                                        "lineHeight": 30.5,
                                        "paddingHorizontal": 18,
                                        "paddingTop": 2,
                                      },
                                      undefined,
                                    ]
                                  }
                                  type="info"
                                >
                                  Copied!
                                </Text>
                                <View
                                  accessibilityLabel="Close toast"
                                  accessibilityRole="button"
                                  accessible={true}
                                  collapsable={false}
                                  focusable={true}
                                  onClick={[Function]}
                                  onResponderGrant={[Function]}
                                  onResponderMove={[Function]}
                                  onResponderRelease={[Function]}
                                  onResponderTerminate={[Function]}
                                  onResponderTerminationRequest={[Function]}
                                  onStartShouldSetResponder={[Function]}
                                  style={
                                    Object {
                                      "opacity": 1,
                                    }
                                  }
                                  testID="toastCloseButton"
                                >
                                  <SvgMock
                                    height={20}
                                    style={
                                      Array [
                                        Object {
                                          "height": 65,
                                          "width": 65,
                                        },
                                        undefined,
                                      ]
                                    }
                                    type="info"
                                    width={20}
                                  />
                                </View>
                              </View>
                            </View>
                          </View>
                        </View>
                      </RCTSafeAreaView>
                    </View>
                  </View>
                </View>
              </View>
            </View>
          </View>
        </RNSScreen>
      </RNSScreenContainer>
      <View
        collapsable={false}
        pointerEvents="box-none"
        style={
          Object {
            "height": 44,
            "left": 0,
            "position": "absolute",
            "right": 0,
            "top": 0,
            "zIndex": 1,
          }
        }
      />
    </View>
  </RNCSafeAreaProvider>
</View>
`;
